include/cuda/memory_resource

```diff
@@ -0,0 +1,973 @@
+ *
+ * Memory allocation kind determines where memory can be accessed and the
+ * performance characteristics of accesses.
+ *
+ * This is not a closed set, the user code can define custom memory kinds.
+ */
+namespace memory_kind {
+  /*!
+   * \brief Ordinary host memory
+   */
+  struct host;
+
+  /*!
+   * \brief Device memory, as allocated by cudaMalloc.
+   */
+  struct device;
+
+  /*!
+   * \brief Device-accessible host memory.
+   */
+  struct pinned;
+
+  /*!
+   * \brief Virtual memory that is automatically migrated between the host and devices.
+   */
+  struct managed;
+};
+
+namespace detail {
+
+template <typename...>
+struct __type_pack {};
+
+namespace __fallback_typeid {
+
+template <class _Tp>
+struct  _LIBCUDACXX_TEMPLATE_VIS __unique_typeinfo { static constexpr int __id = 0; };
+template <class _Tp> constexpr int __unique_typeinfo<_Tp>::__id;
+
+template <class _Tp>
+inline _LIBCUDACXX_INLINE_VISIBILITY
+constexpr const void* __get_fallback_typeid() {
+    return &__unique_typeinfo<std::decay_t<_Tp>>::__id;
+}
+
+template <typename _Tp>
+const ::std::type_info *__get_typeid() {
+#ifdef _LIBCUDACXX_EXT_RTTI_ENABLED
+  return &typeid(_Tp);
+#else
+  return nullptr;
+#endif
+}
+
+inline bool __compare_type(const ::std::type_info *__ti1, const void *__fallback_ti1,
+                           const ::std::type_info *__ti2, const void *__fallback_ti2) {
+#ifdef _LIBCUDACXX_EXT_RTTI_ENABLED
+  if (__ti1 && __ti2 && *__ti1 == *__ti2)
+    return true;
+#endif
+  return __fallback_ti1 == __fallback_ti2;
+}
+
+template <typename _Tp>
+bool __is_type(const ::std::type_info *__ti1, const void *__fallback_ti1) {
+  return __compare_type(__ti1, __fallback_ti1, __get_typeid<_Tp>(), __get_fallback_typeid<_Tp>());
+}
+
+}  // namespace __fallback_typeid
+
+}  // namespace detail
+
+template <typename _ResourcePointer, typename... _Properties>
+class basic_resource_view;
+
+template <typename... _Properties>
+struct properties;
+
+
+namespace detail {
+  template <typename... _Properties>
+  struct __kind_from_properties_helper {
+    using type = void;
+  };
+
+  template <typename... _Properties>
+  using __kind_from_properties_helper_t = typename __kind_from_properties_helper<_Properties...>::type;
+
+  template <typename _Kind, typename... _Tail>
+  struct __kind_from_properties_helper<is_kind<_Kind>, _Tail...> {
+    using type = _Kind;
+  };
+
+  template <typename _FirstProp, typename... _Tail>
+  struct __kind_from_properties_helper<_FirstProp, _Tail...> {
+    using type = __kind_from_properties_helper_t<_Tail...>;
+  };
+}
+
+
+/*!
+ * \brief Evaluaes to the a memory kind tag form _Properties (if listed as is_kind)
+ *        or to a list of properties otherwise.
+ */
+template <typename... _Properties>
+struct kind_from_properties {
+  using type = _CUDA_VSTD::conditional_t<
+    _CUDA_VSTD::is_same<detail::__kind_from_properties_helper_t<_Properties...>, void>,
+    detail::__kind_from_properties_helper_t<_Properties...>
+  >;
```

Comment:
Not sure how this is supposed to compile. `conditional_t` takes 3 template parameters, not two, and the first one is a non-type parameter (so use `is_same_v`). Unfortunately I don't know what the `T` parameter for `conditional_t` should be (the case where the properties list is void). Should it be void as well?:

```suggestion
struct kind_from_properties {
  using type = _CUDA_VSTD::conditional_t<
    _CUDA_VSTD::is_same_v<detail::__kind_from_properties_helper_t<_Properties...>, void>,
    void,
    detail::__kind_from_properties_helper_t<_Properties...>
  >;
```

Reply:
It compiles if I write it as above, but I'm not sure if it's correct.